book-1-orientation/chapters/HTTPCLIENT.md

```diff
@@ -0,0 +1,153 @@
+
+```json
+{
+    "current_page": 1,
+    "limit": 20,
+    "next_page": 1,
+    "previous_page": 1,
+    "results": [
+        {
+            "id": "haMJRfF6hFd",
+            "joke": "How do you fix a broken pizza? With tomato paste."
+        },
+        {
+            "id": "xc21Lmbxcib",
+            "joke": "How did the hipster burn the roof of his mouth? He ate the pizza before it was cool."
+        },
+        {
+            "id": "51DAA5Tfaxc",
+            "joke": "What did Romans use to cut pizza before the rolling cutter was invented? Lil Caesars"
+        },
+        {
+            "id": "rc2E6EdiNe",
+            "joke": "Want to hear my pizza joke? Never mind, it's too cheesy."
+        }
+    ],
+    "search_term": "pizza",
+    "status": 200,
+    "total_jokes": 4,
+    "total_pages": 1
+}
+```
+
+Given the shape of this data, create a new C# class (or classes) that matches these properties.
+
+```csharp
+public class JokeReponse
+{
+    [JsonProperty("current_page")]
+    public int CurrentPage { get; set; }
+
+    public int Limit { get; set; }
+
+    [JsonProperty("next_page")]
+    public int NextPage { get; set; }
+
+    [JsonProperty("previous_page")]
+    public int PreviousPage { get; set; }
+
+    public List<JokeResult> Results { get; set; }
+
+    [JsonProperty("search_term")]
+    public string SearchTerm { get; set; }
+
+    [JsonProperty("status")]
+    public int Status { get; set; }
+
+    [JsonProperty("total_jokes")]
+    public int TotalJokes { get; set; }
+
+    [JsonProperty("total_pages")]
+    public int TotalPages { get; set; }
+}
+
+public class JokeResult
+{
+    public string Id { get; set; }
+    public string Joke { get; set; }
+}
+```
+
+A couple notes about these classes:
+
+- If there is data in the JSON response you don't care about, you do not have to include it in your C# class. For example, if you did not care about the `status` in the previous JSON response, you could leave out the `Status` property in the `JokeResponse` class.
+
+- Not all API's use the same naming conventions. One might use camelCase (i.e. `currentPage`) while another might use snake_case (i.e. `current_page`). By default, the .NET HttpClient is set format responses that use camelCase. If you encounter an API like this one--one that uses a different naming convention--you can handle this by adding a `JsonProperty` annotation above your property names.
+
+```csharp
+[JsonProperty("total_jokes")]
+public int TotalJokes { get; set; }
+```
+
+If creating classes for API responses seems like a tedious process, there are online tools available to help you. Sites like [QuickType](https://quicktype.io/) allow you to paste in an example of your json response and have your C# classes generated for you.
+
```

Comment:
Visual studio does this. https://dailydotnettips.com/did-you-know-you-can-automatically-create-classes-from-json-or-xml-in-visual-studio/